### PR TITLE
[LLM Runtime] Add jblas split weight interface and support jblas models

### PR DESCRIPTION
**Pull request by ClarkChin08**
add jblas tp support

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 1]
```diff
@@ -614,6 +615,124 @@ struct model_model_loader {
...
   }
 
+  void jblas_split_weight(void* src, void** dst, size_t column, size_t row, size_t column_rank, size_t row_rank) {
+    auto src_tmp = jblas::prologue::weight_comp::gemm_kblcok::PackedWeightParser::deserialBuffer(src);
+    // TODO adapt NTILE and KTILE from CoreType
+    int NTILE = 48;
+    int KTILE = 4;
+    int world_size_n = static_cast<int>(src_tmp->mNPad / row);
+    int world_size_k = static_cast<int>(src_tmp->mKPad / column);
+    auto dNpad = src_tmp->mNPad / world_size_n;
```
Comment (luoyu-intel) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:625]: it seems dNpad equals row.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 2]
```diff
@@ -614,6 +615,124 @@ struct model_model_loader {
...
+    auto dKpad = src_tmp->mKPad / world_size_k;
+    // dst NPad and KPad should be div by NTILE and KTILE
+    assert(dNpad / NTILE == 0);
```
Comment (luoyu-intel) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:628]: mod or divide?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 3]
```diff
@@ -614,6 +615,124 @@ struct model_model_loader {
...
+    assert(dKpad / KTILE == 0);
+    if (src_tmp != nullptr) {
+      if (src_tmp->mPrologueID == int(ne_jblas::WeightCompType::WeightS4ClipScaleFp32) ||
+          src_tmp->mPrologueID == int(ne_jblas::WeightCompType::WeightS4ClipScaleFp32PerChannelN)) {
+        auto src_w = dynamic_cast<ne_jblas::SS4Fp32*>(src_tmp); 
+        ne_jblas::SS4Fp32 dst_w(src_w->mCoreType);
+        dst_w.resize(dNpad, dKpad, src_w->mBlockSize, src_w->mIsAsym);
+        dst_w.assign((int8_t*)(*dst));
```
Comment (luoyu-intel) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:636]: what's the size of *dst?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 4]
```diff
@@ -614,6 +615,124 @@ struct model_model_loader {
...
+          auto s_off_ptr = (float*)src_w->mSPtr + i * src_w->mCStep + kblock_offset;
+          auto d_off_ptr = (float*)dst_w.mSPtr + d_row_id * dst_w.mCStep; 
+          memcpy(s_off_ptr, d_off_ptr, dst_w.mCStep);
+          d_row_id += 1;
+        }
+        // take the zp out and split
+        if (src_w->mIsAsym) {
+          d_row_id = 0;
+          for (size_t i = row_rank * src_w->mNPad; i < (row_rank + 1) * src_w->mNPad; ++i) {
+            size_t kblock_offset = column_rank * src_w->mCSize / src_w->mCStep;
```
Comment (luoyu-intel) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:662]: I don't understand this equation. kblock_offset should response row_rank.

**Review by zhewang1-intc — COMMENTED** [review]
how can we quickly verify that this split weight can do gemm correctly on multiple sockets?
Is there a ut for jblas tp?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 4 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:662]: if row_rank responses to N,  what does this 'row_rank * src_w->mNPad' mean?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 5]
```diff
@@ -661,8 +780,14 @@ struct model_model_loader {
       file.seek(shard.file_off, SEEK_SET);
       tmp_buf.resize(lt.size * lt.world_size);
       file.read_raw(tmp_buf.addr, lt.size * lt.world_size);
-      // only copy part of weight form the tmp_buf of origin file
-      memcpy(lt.data, tmp_buf.addr + lt.rank * lt.size, lt.size);
+      size_t num_rows = lt.ne.size() == 1 ? 1 : lt.ne.at(1);
+      if (lt.type == NE_TYPE_JBLAS) {
+        void* dst_data = (void*)lt.data;
+        jblas_split_weight(tmp_buf.addr, &dst_data, lt.ne.at(0), num_rows, lt.rank, 0);
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:786]: Why both `TP_1D_ROW` and `TP_1D_COLUMN` call `jblas_split_weight(tmp_buf.addr, &dst_data, lt.ne.at(0), lt.ne.at(1), lt.rank, 0);`? I assume that they should be slightly different.

**Comment by airMeng** [comment]
> > how can we quickly verify that this split weight can do gemm correctly on multiple sockets? Is there a ut for jblas tp?
> 
> yes, i am verifying on the gptj and llama model. will dump the output and check the accuracy from jblas gemm and then have a performance table later.

any updates?

**Review by luoyu-intel — APPROVED** [review]

**Review by DDEle — APPROVED** [review]

**Review by zhewang1-intc — APPROVED** [review]

**Pull request merged**